### ISSUE TITLE
Unable to ignore SSL certificate validation, request in curl works fine

### ISSUE THREAD

**Issue by kEpEx** [dotnet/runtime#34858]
I am getting the exception bellow when calling a specific HTTPS url, it works for others and same url works fine with curl and other clients.

Our system setup:

OS: Ubuntu 16.04 amd64
Dotnet Core Version: 3.1.100

Curl in same server for same exact URL returns fine. 
Note that the callback and the exception triggers after around 100 seconds, like a client timeout it seems

I am aware of the possible solution by signing the certificate again (#30242) but I don't have control over this. Plus it doesn't explain why curl accepts the certificate and I wonder if there is a way to complete avoid the check of the certificate, which seems like the callaback returning true should but is not working

I also tried the `DangerousAcceptAnyServerCertificateValidator` in the callback but no difference.

```
    $ dotnet run
    4/11/2020 2:05:55 PM Calling URL: https://URL
    4/11/2020 2:07:37 PM Validating Cert
    Unhandled exception. System.Threading.Tasks.TaskCanceledException: The operation was canceled.
        at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
        at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean allowHttp2, CancellationToken cancellationToken)
        at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
        at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
        at testapp.Program.Main(String[] args) in /home/kepex/project/Program.cs:line 23
        at testapp.Program.<Main>(String[] args)    
```

```c#
    using System;
    using System.Threading.Tasks;
    using System.Net.Http;

    namespace testapp
    {
        class Program
        {
            static async Task Main(string[] args)
            {
                var url = "https://URL";
                var handler = new HttpClientHandler()
                {
                    ServerCertificateCustomValidationCallback = (message, cert, chain, sslPolicyErrors) => {  Console.WriteLine(DateTime.Now.ToString() + " Validating Cert"); return true; }
                };
                var client = new HttpClient(handler);
                Console.WriteLine(DateTime.Now.ToString() + " Calling URL: " + url);
                var res = await client.GetAsync(url);
                Console.WriteLine(DateTime.Now.ToString() +" Response Status Code:" + res.StatusCode);
            }
        }
    }
```

**Comment by wfurt** [comment]
curl on Ubuntu 16 .04 uses GnuTLS instead of OpenSSL so validation restrictions will be different. You can check with `curl --version`. However, with custom callback you should be able to bypass validation and you should be able to connect even to untrusted servers.  

Can you post output from `openssl s_client -connect sjc04p1galap01.teslamotors.com:8904`

**Comment by kEpEx** [comment]
Hi the output is bellow,  @wfurt  can you please kindly edit out the servername, I would prefer to keep it private I forgot to remove it earlier.

curl --version output

    curl 7.47.0 (x86_64-pc-linux-gnu) libcurl/7.47.0 GnuTLS/3.4.10 zlib/1.2.8 libidn/1.32 librtmp/2.3
    Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp
    Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP UnixSockets

openssl output:

    openssl s_client -connect server:8904
    CONNECTED(00000003)
    depth=1 C = US, O = "Company, Inc.", OU = PKI, CN = Corporate NA Device Issuing CA 01
    verify error:num=20:unable to get local issuer certificate
    ---
    Certificate chain
    0 s:/CN=servername.domain.com
    i:/C=US/O=Company, Inc./OU=PKI/CN=Corporate NA Device Issuing CA 01
    1 s:/C=US/O=Company, Inc./OU=PKI/CN=Corporate NA Device Issuing CA 01
    i:/C=US/O=Company, Inc./OU=PKI/CN=Corporate Root CA
    ---
    Server certificate
    -----BEGIN CERTIFICATE-----
    MIIFgzCCBGugAwIBAgITPgADH5Pb8wbqLG2gWgAAAAMfkzANBgkqhkiG9w0BAQsF
    ADBdMQswCQYDVQQGEwJVUzEUMBIGA1UEChMLVGVzbGEsIEluYy4xDDAKBgNVBAsT
    A1BLSTEqMCgGA1UEAxMhQ29ycG9yYXRlIE5BIERldmljZSBJc3N1aW5nIENBIDAx
    MB4XDTIwMDIxOTA1MDY1NVoXDTIyMDIxODA1MDY1NVowKTEnMCUGA1UEAxMec2pj
    MDRwMWdhbGFwMDEudGVzbGFtb3RvcnMuY29tMIIBIjANBgkqhkiG9w0BAQEFAAOC
    AQ8AMIIBCgKCAQEAyMb7TaB47WieEHG08K4HBijQumfASXwHYZcxWCWbCwkzxAIn
    9QR5yYmTPJpddK+KpF8+zY1SnejKLLj6Esc1AGlwoTiZsCCugaYf5J/kbmc+h218
    pFDSoYaVO99L2ckoh8srFuBdD8GTC/dwHCMXPErmyUYHSLmO+gb+IwX6ohcdSuzR
    XAF74lJlK4VyGTs/aQkyDKkRji8srJzmj8xnLoIkbHl+gm3FXERLiAeMQzZySudB
    4fZ2rCmXmON9938fWqSjQGOIFFklHF7R2ewegI9awBc/SgXNyx4WHwMlzpbPna2U
    mx0Lm9s79piyLqcneZsDze3xSMjp9Ufl6KiquwIDAQABo4ICbjCCAmowPQYJKwYB
    BAGCNxUHBDAwLgYmKwYBBAGCNxUIh+mEQ4WYuEDdmzeHkbkTg+emQIEtg/exBoPv
    my4CAWQCAQwwEwYDVR0lBAwwCgYIKwYBBQUHAwEwCwYDVR0PBAQDAgWgMFAGA1Ud
    IARJMEcwRQYLKwYBBAGDgH8CAQIwNjA0BggrBgEFBQcCARYoaHR0cDovL3BraS50
    ZXNsYS5jb20vcG9saWNpZXMvYmFzaWMetcetcetc...
    -----END CERTIFICATE-----
    subject=/CN=servername.domain.com
    issuer=/C=US/O=Company, Inc./OU=PKI/CN=Corporate NA Device Issuing CA 01
    ---
    No client certificate CA names sent
    Client Certificate Types: RSA sign, DSA sign, ECDSA sign
    Requested Signature Algorithms: RSA+SHA256:RSA+SHA384:RSA+SHA1:ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA1:DSA+SHA1:RSA+SHA512:ECDSA+SHA512
    Shared Requested Signature Algorithms: RSA+SHA256:RSA+SHA384:RSA+SHA1:ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA1:DSA+SHA1:RSA+SHA512:ECDSA+SHA512
    Peer signing digest: SHA256
    Server Temp Key: ECDH, P-256, 256 bits
    ---
    SSL handshake has read 3380 bytes and written 443 bytes
    ---
    New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-GCM-SHA384
    Server public key is 2048 bit
    Secure Renegotiation IS supported
    Compression: NONE
    Expansion: NONE
    No ALPN negotiated
    SSL-Session:
        Protocol  : TLSv1.2
        Cipher    : ECDHE-RSA-AES256-GCM-SHA384
        Session-ID: 1A1D000087CFF935AF658E692192AA4E673CF54DF9E11A7231F3B31C8666FETC
        Session-ID-ctx:
        Master-Key: 3D56E52DCB1B279DF7C373E5D2F5A6C6DB4F0232FC9AC8A9D84341A38C869B50537A2144C476BF4169A0CCDD05524ETC
        Key-Arg   : None
        PSK identity: None
        PSK identity hint: None
        SRP username: None
        Start Time: 1586665910
        Timeout   : 300 (sec)
        Verify return code: 20 (unable to get local issuer certificate)
    ---

    HTTP/1.1 400 Bad Request
    Connection: close
    Date: Sun, 12 Apr 2020 04:31:51 GMT
    Server: Kestrel
    Content-Length: 0

    read:errno=0

**Comment by wfurt** [comment]
Few thoughts on this @kEpEx. It is curious that the OpenSSL claims unknown CA. Maybe that is expiated few other people hit this. You can run 'openssl version -a' and check OPENSSLDIR to see where OpenSSL expects trusted CA certificates. That may be a link but make sure all the data are readable bu user you running as.  You can check https://askubuntu.com/questions/73287/how-do-i-install-a-root-certificate or others to add it and run `sudo update-ca-certificates`.


Now, I'm wondering if this is TLS problem at all. You can can get `The operation was canceled` when timeout happens and 100s is default Timeout for HttpClient https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.timeout?view=netcore-3.1
That would generally mean the server is not sending data back. You can sort of see that with Wireshark, even if you do not decrypt. I start wondering if you actually hitting some symptoms described in  #805. Some servers from the list did not responded back at all without client sending certain headers in the request. Curl sends some by default (you would see that with curl -v) but HttpClient does not. You may try to add the User-Agent and Accept header.

**Comment by kEpEx** [comment]
I added the Accept and User-Agent that I see in the `curl -v` but still same output


    ...
    Console.WriteLine(DateTime.Now.ToString() + " Calling URL: " + url);
    client.DefaultRequestHeaders.Add("User-Agent", "curl/7.47.0");
    client.DefaultRequestHeaders.Add("Accept", "*/*");

    Console.WriteLine(DateTime.Now.ToString() +" " +client.DefaultRequestHeaders.Accept.ToString());
    Console.WriteLine(DateTime.Now.ToString() +" " +client.DefaultRequestHeaders.UserAgent.ToString());
    ...

Output:


    4/13/2020 11:54:33 AM Calling URL: https://URL
    4/13/2020 11:54:33 AM */*
    4/13/2020 11:54:33 AM curl/7.47.0
    4/13/2020 11:56:14 AM Validating Cert
    Unhandled exception. System.Threading.Tasks.TaskCanceledException: The operation was canceled.
        at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
        at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean allowHttp2, CancellationToken cancellationToken)
        at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
        at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
        at testapp.Program.Main(String[] args) in /home/kepex/project/Program.cs:line 24
        at testapp.Program.<Main>(String[] args)

I will explore the openssl aspect of it, but still wonder how to bypass cert validation in the meantime doesn't work

**Comment by kEpEx** [comment]
Additionally tried curl removing accept and user agent with: `curl -v -H'Accept: ' -H 'User-Agent: '  'https://URL'`

Previous output:

    * ALPN, server did not agree to a protocol
    > GET /PATH HTTP/1.1
    > Host: SERVER:8904
    > User-Agent: curl/7.47.0
    > Accept: */*
    >
    < HTTP/1.1 401 Unauthorized
    < Date: Mon, 13 Apr 2020 19:14:51 GMT
    < Server: Kestrel
    < Content-Length: 0
    < Cache-Control: no-cache

Still work without sending Accept/Agent:

    * ALPN, server did not agree to a protocol
    > GET /PATH HTTP/1.1
    > Host: SERVER:8904
    >
    < HTTP/1.1 401 Unauthorized
    < Date: Mon, 13 Apr 2020 19:13:30 GMT
    < Server: Kestrel
    < Content-Length: 0
    < Cache-Control: no-cache

**Comment by TechnologyAnimal** [comment]
Not sure if it's helpful at all since this is in powershell, but I was able to ignore invalid certificates with System.Net.Security.SslStream on line 229 from the link below.

https://github.com/TechnologyAnimal/Test-TlsProtocols/blob/master/Test-TlsProtocols/Test-TlsProtocols.psm1

**Comment by wfurt** [comment]
The problem is that this is _not_ complaining about certificate but it is probably timing out @TechnologyAnimal. You would see authentication exception and complain about certificate.

**Comment by wfurt** [comment]
I'm generally out of idea @kEpEx. Since this is happening only inside of your private network this will be difficult to resolve. One more test you can try is to connect directly using SslStream. Use AuthenticateAsClientAsync() and then write RAW HTTP request and get response back. If the authenticate call fails, it would be proof that there is something wrong with the handshake. 

Since you can reproduce it with trivial program, it would be interesting to try it with https://dotnet.microsoft.com/download/dotnet/5.0 I'm not aware of any particular change but the SslStream went through major refactor and it may yield some useful info.

**Comment by alnikola** [comment]
Triage:
@wfurt Could you please comment on this and confirm whether it's an issue in our code?

**Comment by wfurt** [comment]
We don't have enough data @alnikola. With no way how to repro at this point, trying SslStream   directly may be best next step.

**Comment by rdasan** [comment]
@wfurt Not exactly in the context of ignoring the SSL, but we are getting timeout errors intermittently in place for 401 in our of our internal API when we use an Invalid Certificate. I saw you mentioned #805 above in one of your comments. Please see my latest comment on that issue. https://github.com/dotnet/runtime/issues/805#issuecomment-633847552

**Comment by wfurt** [comment]
#805 was specific to macOS. What OS do you have @rdasan ?  In cases of generic errors, the reasons and condition can be very different and it does not make sense IMHO to link them just because end result is similar unless the pattern match as well. The linked issues was also very consistent e.g. specific to particular sites. 
If you can make it fail in reasonable time, you may try it with SslStream directly. At this point it is not clear if this is problem with SslStream or if there is something at wrong at HTTP layer. 
Can you check on the server if the HTTP request was received? If you have packet capture it may be possible to guess from TLS change. (e.g. you would see application data after handshake)

**Comment by rdasan** [comment]
Since the other issue is closed  (#805) i will try to give more context here. 
@wfurt We deploy our service as a Web App on Azure App Services in our CI/CD Staging release. The API end points on our service require a Certificate over Https. We have an automated Integration test in our CI/CD pipeline that we run across the our end points where we puposely pass in an Invalid Certificate and in the Test assertion check SPECIFICALLY for a 401 or 403 (Azure App Services return a 403). But at times (intermittently), we get a timeout error with the following exception and our Release(Deployment) fails for that 1 partcular ITest. All the other ITests where we pass in Valid Certificate pass. Only this one fails intermittently.
```
System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.IO.IOException: Unable to read data from the transport connection: Operation timed out.
 ---> System.Net.Sockets.SocketException(60): Operation timed out
   --- End of inner exception stack trace ---
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.GetResult(Int16 token)
   at System.Net.Security.SslStream.<FillBufferAsync>g__InternalFillBufferAsync|215_0[TReadAdapter](TReadAdapter adap, ValueTask`1 task, Int32 min, Int32 initial)
   at System.Net.Security.SslStream.ReadAsyncInternal[TReadAdapter](TReadAdapter adapter, Memory`1 buffer)
   at System.Net.Http.HttpConnection.FillAsync()
   at System.Net.Http.HttpConnection.ReadNextResponseHeaderLineAsync(Boolean foldedHeadersAllowed)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync(HttpConnection connection, HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
```

Hope that helps you understand the issue that we are facing in a better way...

**Comment by wfurt** [comment]
From the call stack it looks like the ssl handshake was completed and HttpClient is trying to read response. If there is issue with handshake you would see stablishSslConnectionAsyncCore() in the trace as in the original post. 
I assume you tried to increate the HttpClient timeout just to be sure?

What version of .NET are you on @rdsan? Would collecting traces be feasible for you? 
Or getting core dump from time when it happened. 

Since I would expect the 40x responses be small, I'm wondering if there is some issue with flushing the response on server side.

**Comment by wfurt** [comment]
do you have any more updates or insights @rdasan? II think getting traces or packet captures could give us some clues.

**Comment by rdasan** [comment]
Sorry @wfurt , got caught up in other projects at work. We already have a pretty liberal timeout sent on the HttpClient. So i don't think timeout is a problem. We are on .Net Core 3.1. The problem with this issue is it happens only intermittently and sometimes does not happen for days together. We have not been able to establish any pattern on when it's happening. One thing we did notice, it never happened on a developer initiated release (merge to develop or merge to master). But we have nightly builds that run on our develop environoment and this issue seems to happen intermittently on our nightly builds alone. I can try to get traces next time it happens. Core dump may not be possible. and we saw this issue only on Azure App Service deployment. In production, ATM this service runs as a Windows Service and so far we have not seen the issue in prod
